src/peft/tuners/delora/layer.py

```diff
@@ -0,0 +1,630 @@
+        An = torch.clamp(A_.norm(dim=1), min=1e-12)
+        Bn = torch.clamp(B_.norm(dim=0), min=1e-12)
+        diag = torch.diag_embed((alpha_ / r) / (An * Bn))
+        delta = B_ @ diag @ A_
+        delta = delta * w_norm_.unsqueeze(0)
+        return delta.to(A.dtype, copy=False)
+
+    def get_delta_weight(self, adapter: str) -> torch.Tensor:
+        if adapter not in self.delora_A or adapter not in self.delora_B:
+            raise ValueError(f"Adapter {adapter} not found.")
+        base_w = self.get_base_layer().weight
+        out_dtype = base_w.dtype
+        device = base_w.device
+        comp_dtype = torch.float32 if out_dtype in (torch.float16, torch.bfloat16) else out_dtype
+
+        A = self.delora_A[adapter].to(dtype=comp_dtype, device=device)
+        B = self.delora_B[adapter].to(dtype=comp_dtype, device=device)
+        alpha = self.delora_alpha[adapter].to(dtype=comp_dtype, device=device)
+        An = torch.clamp(A.norm(dim=1), min=1e-12)
+        Bn = torch.clamp(B.norm(dim=0), min=1e-12)
+        diag = torch.diag_embed((alpha / self.r[adapter]) / (An * Bn))  # (r,r)
+        delta = B @ diag @ A  # (out,in)
+
+        if (not self.use_residual_init) and (adapter in self.delora_initial_A and adapter in self.delora_initial_B):
+            A0 = self.delora_initial_A[adapter].to(dtype=comp_dtype, device=device)
+            B0 = self.delora_initial_B[adapter].to(dtype=comp_dtype, device=device)
+            alpha0 = self.delora_initial_alpha[adapter].to(dtype=comp_dtype, device=device).squeeze(0)
+            A0n = torch.clamp(A0.norm(dim=1), min=1e-12)
+            B0n = torch.clamp(B0.norm(dim=0), min=1e-12)
+            diag0 = torch.diag_embed((alpha0 / self.r[adapter]) / (A0n * B0n))
+            delta = delta - (B0 @ diag0 @ A0)
+        # scale by a fixed per-adapter reference base weight norm when available (avoids caching for exact unmerge)
+        if adapter in self.delora_w_norm:
+            w_norm = self.delora_w_norm[adapter].to(dtype=comp_dtype, device=device)
+        else:
+            # fallback: compute current base weight norm on-the-fly
+            w = base_w
+            w_norm = torch.norm(w.data.to(comp_dtype), dim=0).detach().to(dtype=comp_dtype, device=device)
+        delta = (delta * w_norm.unsqueeze(0)).to(out_dtype)
+        return delta
+
+    def update_layer(
+        self,
+        adapter_name: str,
+        r: int,
+        alpha: float,
+        module_dropout: float,
+        init_weights: bool | Literal["kaiming_uniform", "gaussian", "xavier_uniform"] = True,
+        use_residual_init: Optional[bool] = None,
+    ) -> None:
+        if r <= 0:
+            raise ValueError(f"`r` should be a positive integer value but the value passed is {r}")
+
+        self.r[adapter_name] = r
+        if use_residual_init is not None:
+            # allow toggling via update for already-wrapped layers
+            self.use_residual_init = bool(use_residual_init)
```

Comment:
I don't think we need casting to `bool`, the argument is documented as being of `bool` type and we generally should assume that users pass the correct type (otherwise, we'd have to carefully check each possible argument).